pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.io.util;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+
+
+public class ValueReaderComparisons {
+
+  private ValueReaderComparisons() {
+  }
+
+  private static int mismatch(PinotDataBuffer dataBuffer, long startOffset, int length, ByteBuffer buffer) {
+    boolean littleEndian = dataBuffer.order() == ByteOrder.LITTLE_ENDIAN;
+    if (littleEndian) {
+      buffer.order(ByteOrder.LITTLE_ENDIAN);
+    }
+    int limit = Math.min(length, buffer.limit());
+    int loopBound = limit & ~0x7;
+    int i = 0;
+    for (; i < loopBound; i += 8) {
+      long ours = dataBuffer.getLong(startOffset + i);
+      long theirs = buffer.getLong(i);
+      if (ours != theirs) {
+        long difference = ours ^ theirs;
+        return i + ((littleEndian ? Long.numberOfTrailingZeros(difference) : Long.numberOfLeadingZeros(difference))
+            >>> 3);
+      }
+    }
+    for (; i < limit; i++) {
+      byte ours = dataBuffer.getByte(startOffset + i);
+      byte theirs = buffer.get(i);
+      if (ours != theirs) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded, byte[] bytes) {
+    // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+    if (mismatchPosition == -1) {
+      if (padded && bytes.length < length) {
+        // need to figure out whether the unpadded string is longer than the parameter or not
+        if (bytes.length == 0) {
```

Comment:
IMO we can simplify this logic and change `LoaderTest` to make it pass. If we want to keep the behavior the same in this PR, we can do a separate PR with just this behavior change.

Reply:
I'm wary of removing test cases, it _looks_ like a regression test, it looks like @siddharthteotia added it so he should probably be consulted before removing it. Let's not do it on this branch though. FWIW I feel like the number of special cases is just the right side of manageable though, and if we can make this distinguish between zero padded strings and their valid prefixes with one special case, I'm personally inclined to do that.